src/common/DatePicker/DatePicker.js

```diff
@@ -2,73 +2,110 @@ import React, {
   decodeLocale,
   formatDate,
   generateCalendar,
+  getDateMask,
+  getDatePipe,
+  getDateRegEx,
   getWeekDays,
   getWeekStart,
   months
 } from '../../utils/datePicker.util'
+import {
+  initialState,
+  datePickerActions,
+  datePickerReducer
+} from './datePickerReducer'
 
 import './datePicker.scss'
 
-const DatePicker = ({ onChange, splitCharacter, value }) => {
-  const [date, setDate] = useState(value || new Date())
+const DatePicker = ({
+  date,
+  dateTo,
+  label,
+  onChange,
+  splitCharacter,
+  type
+}) => {
+  const [datePickerState, datePickerDispatch] = useReducer(
+    datePickerReducer,
+    initialState
+  )
   const [isTopPosition, setIsTopPosition] = useState(false)
   const [openDatePicker, setOpenDatePicker] = useState(false)
-  const [selectedDate, setSelectedDate] = useState(value || new Date())
+  const [isRange] = useState(type.includes('range'))
+  const [isTime] = useState(type.includes('time'))
+  const [isCalendarInvalid, setIsCalendarInvalid] = useState(false)
   const [valueDatePickerInput, setValueDatePickerInput] = useState(
-    formatDate(value, splitCharacter)
+    formatDate(isRange, isTime, splitCharacter, date, dateTo)
   )
 
   const datePickerRef = React.createRef()
   const datePickerViewRef = useRef()
-  const maskDate = [
-    /\d/,
-    /\d/,
-    splitCharacter,
-    /\d/,
-    /\d/,
-    splitCharacter,
-    /\d/,
-    /\d/,
-    /\d/,
-    /\d/
-  ]
-  const autoCorrectedDatePipe = createAutoCorrectedDatePipe('mm/dd/yyyy')
+  const dateMask = getDateMask(isRange, isTime, splitCharacter)
+  const autoCorrectedDatePipe = createAutoCorrectedDatePipe(
+    getDatePipe(isRange, isTime),
+    undefined,
+    isRange && datesDivider
+  )
+  const dateRegEx = getDateRegEx(dateMask)
   const startWeek = getWeekStart(decodeLocale(navigator.language))
 
-  const calendar = useMemo(() => {
-    return generateCalendar(date || new Date(), startWeek)
-  }, [date, startWeek])
-
   const handleCloseDatePickerOutside = useCallback(
     event => {
       if (
         datePickerRef.current &&
         !datePickerRef.current.contains(event.target)
       ) {
+        datePickerDispatch({
+          type: datePickerActions.UPDATE_VISIBLE_DATE_FROM,
+          payload: datePickerState.configFrom.date
+        })
+
+        if (isRange) {
+          datePickerDispatch({
+            type: datePickerActions.UPDATE_VISIBLE_DATE_TO,
+            payload: datePickerState.configTo.date
+          })
+        }
+
         setOpenDatePicker(false)
-        setDate(selectedDate || new Date())
       }
     },
-    [datePickerRef, selectedDate]
+    [
+      datePickerRef,
+      datePickerState.configFrom.date,
+      datePickerState.configTo.date,
+      isRange
+    ]
   )
 
+  const isSameDate = (firstDate, secondDate) => {
+    return (
+      firstDate.getDate() === secondDate.getDate() &&
+      firstDate.getMonth() === secondDate.getMonth() &&
+      firstDate.getFullYear() === secondDate.getFullYear()
```

Comment:
Do not change this code!
But I'm just sharing these ideas to match dates that came to my mind when seeing the above code:
```js
firstDate.toISOString().slice(0, 10) === secondDate.toISOString().slice(0, 10)

firstDate.toISOString().split('T')[0] === secondDate.toISOString().split('T')[0]
```
Of course, they're much less readable, so don't use them :)